articles/what-is-an-abi/index.md

```diff
@@ -0,0 +1,268 @@
+
+- If the type-kind of the return type is a Float then the register `xmm0` will be used. If two registers are needed then `xmm1` will be the second register.
+
+- If the type-kind of the return type is a Pointer then treat it as an Integer.
+
+- If the type-kind of the return type is an Aggregate and the size can fit into two registers then follow the steps for the sub-type-kind (Pointer, Integers, or Float only). If it is too big or the sub-type-kind is Memory then the caller will need to allocate space for it and pass the pointer to said space as a Pointer type in `rdi` (as if it was the first argument). Then the callee will return the same pointer as a Pointer type in `rax`.
+
+A lot is going on here so let's do some examples:
+
+First, we need to define some structs:
+```c++
+typedef struct { int32_t a; int32_t b; } TwoInts;
+typedef struct { int32_t a; int32_t b; int32_t c; } ThreeInts;
+typedef struct { int32_t a; int32_t b; int32_t c; int32_t d; } FourInts;
+typedef struct { int32_t a; int32_t b; int32_t c; int32_t d; int32_t e; } FiveInts;
+```
+
+With this function signature we get:
+```c++
+FiveInts test_func1(int a, float b, TwoInts c, ThreeInts d);
+```
+->
+```
+var             type                where
+return                 FiveInts*    rdi
+a                        int32_t    rsi
+b                          float    xmm0
+c             {int32_t, int32_t}    rdx
+d    {int32_t, int32_t, int32_t}    rcx, r8
+-------------------------------------------
+return                 FiveInts*    rax (points to stack)
+```
+
+With this function signature we get:
+```c++
+FourInts test_func4(int a, int b, int c, int d)
+```
+->
+```
+var                                  type       where
+a                                    int32_t    rdi
+b                                    int32_t    rsi
+c                                    int32_t    rdx
+d                                    int32_t    rcx
+-----------------------------------------------------
+return  {int32_t, int32_t, int32_t, int32_t}    rax, rdx
+```
+
+I made a lot more "interactive" examples [here](https://repl.it/@ZackJorquera/c-calling-conventions).
+
+Now, it should go without saying that, just like for structs, Rust does this a little differently. Because this article is getting a little long I won't go over it in depth. 
+
+Also a little side-note, rust currently [doesn't have a stable ABI](https://github.com/rust-lang/rfcs/issues/600) so everything is subject to change (however not often). This, as far as rust is concerned, isn't necessarily a bad thing when it comes to writing production code as it really only means that using two different versions of the rust compiler might result in backward compatibility issues. This in practice will never happen because [cargo](https://doc.rust-lang.org/cargo/) will compile all your dependencies for you together.

```

Comment:
as rust is concerned -> as Rust is concerned

Reply:
CAPITALIZE Rust (unless it's generally stylized as rust, in which case, don't have rust with a capital)